packages/vincent-tools/vincent-tool-sdk/src/lib/get-onchain-policy-params.ts

```diff
@@ -0,0 +1,97 @@
+                // to validate that there are no required Policy params
+                return parseOnChainPolicyParams(args.policyUserParamsSchema.parse({}));
+            } else {
+                const decodedPolicyParams = abiDecodePolicyParameters({ params: onChainPolicyParams.parameters });
+                return parseOnChainPolicyParams(args.policyUserParamsSchema.parse(decodedPolicyParams));
+            }
+        } else {
+            if (!onChainPolicyParams) {
+                return;
+            } else {
+                throw new Error(`Agent Wallet PKP Token ID: ${args.agentWalletPkpTokenId} has registered on-chain Policy parameters for Vincent App: ${allOnChainPolicyParams.appId.toString()} App Version: ${allOnChainPolicyParams.appVersion.toString()} Vincent Policy: ${args.policyIpfsCid} Vincent Tool: ${args.toolIpfsCid} but no userParamsSchema was defined by the Policy`);
+            }
+        }
+    } catch (error) {
+        throw new Error(`Error getting on-chain policy parameters (getOnChainPolicyParams): ${error instanceof Error ? error.message : String(error)}`);
+    }
+}
+
+const _getAllOnChainPolicyParams = async (args: {
+    yellowstoneRpcUrl: string,
+    vincentContractAddress: string,
+    appDelegateeAddress: string,
+    agentWalletPkpTokenId: string,
+    toolIpfsCid: string,
+}): Promise<AllOnChainPolicyParams> => {
+    try {
+        const VINCENT_CONTRACT_ABI = [
+            `function validateToolExecutionAndGetPolicies(address delegatee, uint256 pkpTokenId, string calldata toolIpfsCid) external view returns (tuple(bool isPermitted, uint256 appId, uint256 appVersion, tuple(string policyIpfsCid, tuple(string name, uint8 paramType, bytes value)[] parameters)[] policies) validation)`,
+        ];
+
+        const vincentContract = new ethers.Contract(
+            args.vincentContractAddress,
+            VINCENT_CONTRACT_ABI,
+            new ethers.providers.JsonRpcProvider(
+                args.yellowstoneRpcUrl
+            )
+        );
+
+        return vincentContract.validateToolExecutionAndGetPolicies(
+            args.appDelegateeAddress,
+            args.agentWalletPkpTokenId,
+            args.toolIpfsCid
+        );
+    } catch (error) {
+        throw new Error(`Error getting on-chain policy parameters from Vincent contract: ${args.vincentContractAddress} using App Delegatee: ${args.appDelegateeAddress} and Agent Wallet PKP Token ID: ${args.agentWalletPkpTokenId} and Vincent Tool: ${args.toolIpfsCid}: ${error instanceof Error ? error.message : String(error)}`);
+    }
+}
+
+const parseOnChainPolicyParams = (onChainPolicyParams: Record<string, EthersAbiDecodedValue>, policyUserParamsSchema?: z.infer<VincentPolicyDef['userParamsSchema']>) => {
+    try {
+        return policyUserParamsSchema.parse(onChainPolicyParams);
+    } catch (error) {
+        const errorMessage = error instanceof z.ZodError ? formatZodError(error) : error instanceof Error ? error.message : String(error);
```

Comment:
formatZodError returns an object w/ message and other data -- we want that object to be converted to an error `message` string for use here - don't just rely on the object being stringified.